contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol

```diff
@@ -0,0 +1,224 @@
+        uint256 ephemeralPubkey,
+        uint8 stealthPubkeyPrefix,
+        uint8 dhkeyPrefix,
+        uint8 ephemeralPrefix
+    ) external returns (address) {
+        if (_smartAccountStealth[msg.sender].stealthAddress != address(0)) {
+            revert AlreadyInitedForSmartAccount(msg.sender);
+        }
+        if (stealthAddress == address(0))
+            revert ZeroAddressNotAllowedAsStealthAddress();
+        _smartAccountStealth[msg.sender] = StealthStorage(
+            stealthPubkey,
+            dhkey,
+            ephemeralPubkey,
+            stealthAddress,
+            stealthPubkeyPrefix,
+            dhkeyPrefix,
+            ephemeralPrefix
+        );
+        return address(this);
+    }
+
+    /// @inheritdoc IAuthorizationModule
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) external view virtual returns (uint256 validationData) {
+        (bytes memory cleanSignature, ) = abi.decode(
+            userOp.signature,
+            (bytes, address)
+        );
+        bytes1 mode = cleanSignature[0];
+        assembly {
+            let len := mload(cleanSignature)
+            mstore(add(cleanSignature, 0x01), sub(len, 1))
+            cleanSignature := add(cleanSignature, 0x01)
+        }
+
+        // 0x00: signature from spending key
+        // 0x01: aggregated signature from owner and shared secret
+        if (mode == 0x00) {
+            if (_verifySignature(userOpHash, cleanSignature, userOp.sender)) {
+                return 0;
+            }
+            return SIG_VALIDATION_FAILED;
+        } else if (mode == 0x01) {
+            if (
+                _verifyAggregateSignature(
+                    userOpHash,
+                    cleanSignature,
+                    userOp.sender
+                )
+            ) return 0;
+        }
+        return SIG_VALIDATION_FAILED;
+    }
+
+    /// @inheritdoc IStealthAddressRegistryModule
+    function getStealthAddress(
+        address smartAccount
+    ) external view returns (StealthStorage memory) {
+        return _smartAccountStealth[smartAccount];
+    }
+
+    /**
+     * @dev Returns the the magic value of EIP-1271.
+     * @param dataHash The hash of the data signed.
+     * @param moduleSignature The signature of the data.
+     * @return magicValue The magic value.
+     */
+    function isValidSignature(
+        bytes32 dataHash,
+        bytes memory moduleSignature
+    ) public view virtual override returns (bytes4) {
+        bytes1 mode = moduleSignature[0];
+        assembly {
+            let len := mload(moduleSignature)
+            mstore(add(moduleSignature, 0x01), sub(len, 1))
+            moduleSignature := add(moduleSignature, 0x01)
+        }
+
+        // 0x00: signature from spending key
+        // 0x01: aggregated signature from owner and shared secret
+        if (mode == 0x00) {
+            if (_verifySignature(dataHash, moduleSignature, msg.sender)) {
```

Comment:
Can two SAs be owned by the same Stealth address?
If yes, it is required to make the resulting hash for 1271 signatures to depend on the SA's address